x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx

```diff
@@ -0,0 +1,210 @@
+import { RouteComponentProps, withRouter } from 'react-router-dom';
+import { updateTimePicker } from 'x-pack/plugins/apm/public/store/urlParams';
+import {
+  fromQuery,
+  getQueryWithRisonParams,
+  risonSafeDecode,
+  toQuery
+} from '../Links/url_helpers';
+
+interface Props extends RouteComponentProps {
+  dispatchUpdateTimePicker: typeof updateTimePicker;
+}
+
+interface State {
+  from: string;
+  to: string;
+  isPaused: boolean;
+  refreshInterval: number;
+}
+
+interface QueryG {
+  time?: {
+    from: string;
+    to: string;
+  };
+  refreshInterval?: {
+    pause: boolean;
+    value: number;
+  };
+}
+
+// TODO move into EUI as PR
+declare module '@elastic/eui' {
+  interface OnTimeChangeProps {
+    start: string;
+    end: string;
+    isInvalid: boolean;
+    isQuickSelection: boolean;
+  }
+
+  interface OnRefreshChangeProps {
+    isPaused: boolean;
+    refreshInterval: number;
+  }
+
+  export type EuiSuperDatePickerProps = CommonProps & {
+    start: string;
+    end: string;
+    isPaused: boolean;
+    refreshInterval: number;
+    onTimeChange: (props: OnTimeChangeProps) => void;
+    onRefreshChange: (props: OnRefreshChangeProps) => void;
+  };
+  export const EuiSuperDatePicker: React.SFC<EuiSuperDatePickerProps>;
+}
+
+const APM_DEFAULT_TIME_RANGE = {
+  from: 'now-24h',
+  to: 'now'
+};
+
+const APM_DEFAULT_REFRESH = {
+  pause: true,
+  value: 0
+};
+
+class DatePickerComponent extends React.Component<Props, State> {
+  public state = {
+    from: '',
+    to: '',
+    isPaused: true,
+    refreshInterval: 0
+  };
+  public lastRefresh = 0;
+  public nextRefreshRafId = 0;
+
+  public componentDidMount() {
+    // read from query string to initialize, or set defaults
+    const g = this.getG(this.props.location.search);
+    this.updateStateFromUrl(g);
+  }
+
+  public componentWillUnmount() {
+    cancelAnimationFrame(this.nextRefreshRafId);
+  }
+
+  public componentDidUpdate(prevProps: Props) {
+    if (prevProps.location.search !== this.props.location.search) {
+      // TODO do we need to check for this
+      const currentG = this.getG(this.props.location.search);
+      const previousG = this.getG(prevProps.location.search);
+      if (currentG !== previousG) {
+        this.updateStateFromUrl(currentG);
+      }
+    }
+  }
+
+  public getG(search: string) {
+    const { _g } = toQuery(search);
+    return risonSafeDecode(_g) as QueryG | undefined;
+  }
+
+  public refresh = (min: string, max: string) => {
+    cancelAnimationFrame(this.nextRefreshRafId);
+    const { isPaused, refreshInterval } = this.state;
+    const now = new Date().getTime();
+
+    if (isPaused || !refreshInterval) {
+      return;
+    }
+    if (!this.lastRefresh) {
+      this.lastRefresh = now;
+    }
+    if (now - this.lastRefresh >= refreshInterval) {
+      this.props.dispatchUpdateTimePicker({ min, max });
+      this.lastRefresh = now;
+    }
+    requestAnimationFrame(() => this.refresh(min, max));
```

Comment:
Why is `raf` needed?

Reply:
@ogupte and I talked about the dangers of using `setInterval` last week, and we were flipping between this and a recursive `setTimeout`. Still thinking about which way I want to go there.

Reply:
Removed raf in favor of recursive setTimeout